docs/adrs/2023-11-02-min-runners-strategy.md

```diff
@@ -0,0 +1,23 @@
+# Introducing min runner strategies
+
+**Status**: Proposed
+
+## Context
+
+Current implementation treats the `minRunners` field as the number of runners that should be running on your cluster. They can be busy running the job, or they can be idle. This ensures faster cold startup time when workflows are acquired as well as trying to use the minimum amount of runners needed to fulfill the scaling requirement.
+
+However, large and busy clusters could benefit having `minRunners` as `minIdleRunners`. When jobs are comming in large batches, the `AutoscalingRunnerSet` should pre-emptively increase the number of idle runners to further decrease the startup time for the next batch. In that scenario, the amount of runners that should be created should be calculated as the number of acquired workflows plus the number of `minRunners`.
+
+## Decision
+
+The decision is to include an additional `minRunnerStrategy` field per `AutoscalingRunnerSet`. This field will allow the scaling strategy to be picked based on the scale set. There will be two strategies: "lazy" and "eager".
+
+With the "lazy" strategy, the current behavior of `minRunners` will be preserved. The `minRunners` field will specify the minimum number of runners running in a cluster regardless of their state ("running" or "idle").
+
+With the "eager" strategy, `minRunners` will be treated as `minIdleRunners`. This strategy will calculate the number of runners based on the number of workflows acquired plus the number of minRunners. If the `maxRunners` field is specified, it will be respected, so the number of idle runners can be less than the number of idle `minRunners` when the number of acquired jobs plus the number of min runners is greater than `maxRunners`.
```

Comment:
since we don't stop acquire jobs after we reaching maxRunners, would this cause we creating too many runners?
also since 2 runner scale sets can acquire the same job, but only one of them will get run, will we cause too much wasting resource?

Reply:
1. It should not because the patch from the listener should only be up to `maxRunners` regardless of how many are acquired
2. Here we can over-provision them if other scale set "steals" the job, but eventually we will scale down on the next message. The unlucky scenario here can be that no events are created after the job is stolen, so we may not scale down to the desired number of min runners. Good point...

Reply:
We might also want to fix the scale down process if we start doing this since it might make the extra scale up/down even worse.